kernel/userspace.c

```diff
@@ -0,0 +1,195 @@
+#endif
+
+/* Stub functions, to be filled in forthcoming patch sets */
+
+static void set_thread_perms(struct _k_object *ko, struct k_thread *thread)
+{
+	ARG_UNUSED(ko);
+	ARG_UNUSED(thread);
+
+	/* STUB */
+}
+
+
+static int test_thread_perms(struct _k_object *ko)
+{
+	ARG_UNUSED(ko);
+
+	/* STUB */
+
+	return 1;
+}
+
+static int _is_thread_user(void)
+{
+	/* STUB */
+
+	return 0;
+}
+
+void k_object_grant_access(void *object, struct k_thread *thread)
+{
+	struct _k_object *ko = _k_object_find(object);
+
+	if (!ko) {
+		if (_is_thread_user()) {
+			printk("granting access	to non-existent kernel object %p\n",
+			       object);
+			k_oops();
+		} else {
+			/* Supervisor threads may at times instantiate objects
+			 * that ignore rules on where they can live. Such
+			 * objects won't ever be usable from userspace, but
+			 * we shouldn't explode.
+			 */
+			return;
+		}
+	}
+
+	/* userspace can't grant access to objects unless it already has
+	 * access to that object
+	 */
+	if (_is_thread_user() && !test_thread_perms(ko)) {
+#ifdef CONFIG_PRINTK
+		printk("insufficient permissions in current thread %p\n",
+		       _current);
+		printk("Cannot grant access to %s %p for thread %p\n",
+		       otype_to_str(ko->type), object, thread);
+#endif
+		k_oops();
+	}
+	set_thread_perms(ko, thread);
+}
+
+
+int _k_object_validate(void *obj, enum k_objects otype, int init)
+{
+	struct _k_object *ko;
+
+	ko = _k_object_find(obj);
+
+	if (!ko || ko->type != otype) {
+#ifdef CONFIG_PRINTK
+		printk("%p is not a %s\n", obj, otype_to_str(otype));
+#endif
+		return -EBADF;
+	}
+
+	/* Uninitialized objects are not owned by anyone. However if an
+	 * object is initialized, and the caller is from userspace, then
+	 * we need to assert that the user thread has sufficient permissions
+	 * to re-initialize.
+	 */
+	if (ko->flags & K_OBJ_FLAG_INITIALIZED && _is_thread_user() &&
+	    !test_thread_perms(ko)) {
+#ifdef CONFIG_PRINTK
+		printk("thread %p does not have permission on %s %p\n",
+		       _current, otype_to_str(otype), obj);
+#endif
+		return -EPERM;
+	}
+
+	/* If we are not initializing an object, and the object is not
+	 * initialized, we should freak out
+	 */
+	if (!init && !(ko->flags & K_OBJ_FLAG_INITIALIZED)) {
+		printk("%p used before initialization\n", obj);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+
+void _k_object_init(void *object)
+{
+	struct _k_object *ko;
+
+	/* By the time we get here, if the caller was from userspace, all the
+	 * necessary checks have been done in _k_object_validate(), which takes
+	 * place before the object is initialized.
+	 *
+	 * This function runs after the object has been initialized and
+	 * finalizes it
+	 */
+
+	ko = _k_object_find(object);
+	if (!ko) {
+		/* Supervisor threads can ignore rules about kernel objects
+		 * and may declare them on stacks, etc. Such objects will never
+		 * be usable from userspace, but we shouldn't explode.
+		 */
+		return;
+	}
+
+	memset(ko->perms, 0, CONFIG_MAX_THREAD_BITS);
```

Comment:
Naming mistake or size goof?  If that config value is in bits, then this is clobbering 8x too much memory.  If it's not, it should be renamed.

Reply:
oops, naming sucks, will fix this